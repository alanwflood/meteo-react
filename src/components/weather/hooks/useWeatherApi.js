import { useState, useEffect } from "react";
import WeatherAPI from "../../../api/weatherApi";
import SetDocumentTheme from "../setTheme";

export default function useWeatherApi(lat, lng) {
  const [loadingState, setLoadingState] = useState({
    loading: false,
    showLoader: false,
  });
  const [weatherData, setWeatherData] = useState({});
  const [theme, setTheme] = useState("light");
  const api = WeatherAPI(lat, lng);

  useEffect(() => {
    setLoadingState({ loading: true, showLoader: false });
    const delayLoader = setTimeout(
      () => setLoadingState({ loading: true, showLoader: true }),
      3000
    );

    api
      .fetchFiveDayWeather(({ data, times }) => {
        setWeatherData(data);
        SetDocumentTheme({ ...times }, (theme) => setTheme(theme));
      })
      .then(() => {
        clearTimeout(delayLoader);
        setLoadingState({ loading: false, showLoader: false });
      });

    return () => {
      clearTimeout(delayLoader);
    };
  }, [lat, lng]);

  return {
    loadingState,
    weatherData,
    theme,
  };
}
