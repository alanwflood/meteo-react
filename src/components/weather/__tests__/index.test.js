import React from "react";
import { render } from "@testing-library/react";
import Weather from "../index";
import useWeatherApi from "../hooks/useWeatherApi";

jest.mock("../hooks/useWeatherApi");
jest.mock("../card", () => {
  return {
    __esModule: true,
    A: true,
    default: () => {
      return <div />;
    },
  };
});

describe("Weather Component", () => {
  it("Shows nothing when loading and showloader is false", () => {
    useWeatherApi.mockReturnValue({
      weatherData: {},
      loadingState: {
        loading: true,
        showLoader: false,
      },
      theme: "light",
    });
    const { container } = render(<Weather lat={0} lng={0} />);
    expect(container).toMatchSnapshot();
  });

  it("Shows a loader when loading", () => {
    useWeatherApi.mockReturnValue({
      weatherData: {},
      loadingState: {
        loading: true,
        showLoader: true,
      },
      theme: "light",
    });
    const { container } = render(<Weather lat={0} lng={0} />);
    expect(container).toMatchSnapshot();
  });

  it("Renders when data returns from api", () => {
    useWeatherApi.mockReturnValue({
      weatherData: {
        key: 1,
        otherKey: 2,
      },
      loadingState: {
        loading: false,
        showLoader: false,
      },
      theme: "light",
    });
    const { container } = render(<Weather lat={0} lng={0} />);
    expect(container).toMatchSnapshot();
  });
});
