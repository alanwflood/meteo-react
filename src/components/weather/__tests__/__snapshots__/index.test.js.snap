// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Component Renders when data returns from api 1`] = `
<div>
  <div />
  <div />
</div>
`;

exports[`Weather Component Shows a loader when loading 1`] = `
<div>
  <div
    class="loader main-loader"
  >
    Loading
  </div>
</div>
`;

exports[`Weather Component Shows nothing when loading and showloader is false 1`] = `<div />`;
